nixos/modules/services/networking/nomad.nix

```diff
@@ -108,24 +108,25 @@ in
         iptables
       ]);
 
-      serviceConfig = {
-        DynamicUser = cfg.dropPrivileges;
-        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
-        ExecStart = "${cfg.package}/bin/nomad agent -config=/etc/nomad.json";
-        KillMode = "process";
-        KillSignal = "SIGINT";
-        LimitNOFILE = 65536;
-        LimitNPROC = "infinity";
-        OOMScoreAdjust = -1000;
-        Restart = "on-failure";
-        RestartSec = 2;
-        # Agrees with the default `data_dir = "/var/lib/nomad"` in `settings` above.
-        StateDirectory = "nomad";
-        TasksMax = "infinity";
-        User = optionalString cfg.dropPrivileges "nomad";
-      } // (optionalAttrs cfg.enableDocker {
-        SupplementaryGroups = "docker"; # space-separated string
-      });
+      serviceConfig = mkMerge [
+        {
+          DynamicUser = cfg.dropPrivileges;
+          ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+          ExecStart = "${cfg.package}/bin/nomad agent -config=/etc/nomad.json";
+          KillMode = "process";
+          KillSignal = "SIGINT";
+          LimitNOFILE = 65536;
+          LimitNPROC = "infinity";
+          OOMScoreAdjust = -1000;
+          Restart = "on-failure";
+          RestartSec = 2;
+          # Agrees with the default `data_dir = "/var/lib/nomad"` in `settings` above.
+          StateDirectory = "nomad";
+          TasksMax = "infinity";
+          User = optionalString cfg.dropPrivileges "nomad";
```

Comment:
You may want to just unconditionally set this for simplicity. I'm not sure what `User = "";` means to `systemd` 🤔

Reply:
It seems like it's allowed: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Credentials

> The user/group name must have at least one character, and at most 31.

I will unconditionally set it to `"nomad"`.

Reply:
It looks like setting `User=` in a systemd unit is valid and will use the default of `root`. That does seem consistent with what's written here https://systemd.io/USER_NAMES/, and _sort of_ consistent with https://www.freedesktop.org/software/systemd/man/systemd.exec.html#User= considering the notion of relaxed rules which doesn't seem to preclude the empty string.

An empty string might also have some kind of systemd-wide meaning like "Use the default when an empty string is encountered", but I couldn't find anything to that effect anywhere.

In any case I don't think we should unconditionally set this to nomad, since nomad is perfectly fine to run as root. Unconditionally setting the user to nomad means users must have statically created the `nomad` user or use the dynamic users feature.

Reply:
@aanderse Curious to hear your thoughts here.

Reply:
You're right @cpcloud. I must have been confused when I wrote that comment. In that case I *think* you should probably switch from `optionalString` on the value to `mkIf` around a block to make things more clear in the generate `.service` file. As in:
```
      serviceConfig = mkMerge [
        {
          ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
          ExecStart = "${cfg.package}/bin/nomad agent -config=/etc/nomad.json" +
            concatMapStrings (path: " -config=${path}") cfg.extraSettingsPaths;
          KillMode = "process";
          KillSignal = "SIGINT";
          LimitNOFILE = 65536;
          LimitNPROC = "infinity";
          OOMScoreAdjust = -1000;
          Restart = "on-failure";
          RestartSec = 2;
          TasksMax = "infinity";
        }
        (mkIf cfg.enableDocker {
          SupplementaryGroups = "docker"; # space-separated string
        })
        (mkIf (cfg.settings.data_dir == "/var/lib/nomad") {
          StateDirectory = "nomad";
        })
        (mkIf cfg.dropPrivileges {
          DynamicUser = true;
          User = "nomad";
        })
      ];
```

Sorry for adding confusion :bow:

Reply:
Thanks, that clears it up. I'll drop `User` altogether.